sdks/python/apache_beam/transforms/userstate_test.py

```diff
@@ -494,9 +496,33 @@ def clear_values(self, bag_state=beam.DoFn.StateParam(SET_STATE)):
            | beam.ParDo(SetStateClearingStatefulDoFn())
            | beam.ParDo(self.record_dofn()))
 
-    self.assertEqual(
-        ['extra'],
-        StatefulDoFnOnDirectRunnerTest.all_records)
+    self.assertEqual(['value2'], StatefulDoFnOnDirectRunnerTest.all_records)
+
+  def test_stateful_set_state_fn_runner(self):
+
+    class SetStateClearingStatefulDoFn(beam.DoFn):
+
+      SET_STATE = SetStateSpec('buffer', FastPrimitivesCoder())
+
+      def process(self,
+                  element,
+                  set_state=beam.DoFn.StateParam(SET_STATE)):
+        _, value = element
+        aggregated_value = 0
+        set_state.add(value)
+        for saved_value in set_state.read():
+          aggregated_value += saved_value
+        yield aggregated_value
```

Comment:
Maybe set a timer and test clear as well?

Reply:
I tried to add this test case: 
```
  def test_stateful_set_state_clean_portably(self):

    class SetStateClearingStatefulDoFn(beam.DoFn):

      SET_STATE = SetStateSpec('buffer', VarIntCoder())
      EMIT_TIMER = TimerSpec('emit_timer', TimeDomain.WATERMARK)
      CLEAR_TIMER = TimerSpec('clear_timer', TimeDomain.WATERMARK)

      def process(self,
                  element,
                  set_state=beam.DoFn.StateParam(SET_STATE),
                  emit_timer=beam.DoFn.TimerParam(EMIT_TIMER),
                  clear_timer=beam.DoFn.TimerParam(CLEAR_TIMER)):
        _, value = element
        set_state.add(value)
        if value == 2:
          clear_timer.set(3)
        elif value == 3:
          emit_timer.set(4)

      @on_timer(EMIT_TIMER)
      def emit_values(self, set_state=beam.DoFn.StateParam(SET_STATE)):
          yield sorted(set_state.read())

      @on_timer(CLEAR_TIMER)
      def clear_values(self, set_state=beam.DoFn.StateParam(SET_STATE)):
        set_state.clear()
        set_state.add(100)

    p = TestPipeline()
    values = p | beam.Create(range(1, 5))
    actual_values = (values
                     | beam.Map(lambda t: window.TimestampedValue(t, t))
                     | beam.WindowInto(window.FixedWindows(1))
                     | beam.ParDo(SetStateClearingStatefulDoFn()))

    assert_that(actual_values, equal_to([100]))

    result = p.run()
    result.wait_until_finish()
```
but it didn't work it complained that timer is not supported.
```
(beam) ➜  python git:(BEAM-7741-implement-setstate) ✗ python -m unittest apache_beam.transforms.userstate_test.StatefulDoFnOnDirectRunnerTest.test_stateful_set_state_clean_portably
WARNING:root:Key coder FastPrimitivesCoder for transform <ParDo(PTransform) label=[ParDo(SetStateClearingStatefulDoFn)]> with stateful DoFn may not be deterministic. This may cause incorrect behavior for complex key types. Consider adding an input type hint for this transform.
E
======================================================================
ERROR: test_stateful_set_state_clean_portably (apache_beam.transforms.userstate_test.StatefulDoFnOnDirectRunnerTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "apache_beam/transforms/userstate_test.py", line 567, in test_stateful_set_state_clean_portably
    result = p.run()
  File "apache_beam/testing/test_pipeline.py", line 107, in run
    else test_runner_api))
  File "apache_beam/pipeline.py", line 406, in run
    self._options).run(False)
  File "apache_beam/pipeline.py", line 419, in run
    return self.runner.run_pipeline(self, self._options)
  File "apache_beam/runners/direct/direct_runner.py", line 128, in run_pipeline
    return runner.run_pipeline(pipeline, options)
  File "apache_beam/runners/portability/fn_api_runner.py", line 319, in run_pipeline
    default_environment=self._default_environment))
  File "apache_beam/runners/portability/fn_api_runner.py", line 323, in run_via_runner_api
    stage_context, stages = self.create_stages(pipeline_proto)
  File "apache_beam/runners/portability/fn_api_runner.py", line 385, in create_stages
    use_state_iterables=self._use_state_iterables)
  File "apache_beam/runners/portability/fn_api_runner_transforms.py", line 489, in create_and_optimize_stages
    stages = list(phase(stages, pipeline_context))
  File "apache_beam/runners/portability/fn_api_runner_transforms.py", line 1168, in inject_timer_pcollections
    raise NotImplementedError('Timers and side inputs.')
NotImplementedError: Timers and side inputs.

----------------------------------------------------------------------
Ran 1 test in 0.207s

FAILED (errors=1)
```
This is a little bit confusing for me because I have used Timer on Flink runner and it works fine. This could be a bug on the portable runner side.

Reply:
I still have to debug this, I might need some help to debug this.

Reply:
Hmm... for some reason it looks like it's assuming you're using a timer and a side input in the same DoFn.

Reply:
@robertwb I spent some time debugging this. It seems like FnApiRunner is running into issues when there are two timers uses. 
```  def test_two_timers_portably(self):

    class TwoTimersDoFn(beam.DoFn):

      EMIT_TIMER = TimerSpec('emit_timer', TimeDomain.WATERMARK)
      CLEAR_TIMER = TimerSpec('clear_timer', TimeDomain.WATERMARK)

      def process(self,
                  element,
                  emit_timer=beam.DoFn.TimerParam(EMIT_TIMER),
                  clear_timer=beam.DoFn.TimerParam(CLEAR_TIMER)
                  ):
        _, value = element
        if value == 2:
          clear_timer.set(3)
          pass
        elif value == 3:
          emit_timer.set(4)
          pass

      @on_timer(EMIT_TIMER)
      def emit_values(self):
        yield 'emit1'

      @on_timer(CLEAR_TIMER)
      def clear_values(self):
        yield 'clear1'

    p = TestPipeline()
    values = p | beam.Create([('key', 1),
                              ('key', 2),
                              ('key', 3),
                              ('key', 4),
                              ('key', 5),
                              ('key', 6)])
    actual_values = (values
                     | beam.Map(lambda t: window.TimestampedValue(t, t[1]))
                     | beam.WindowInto(window.FixedWindows(1))
                     | beam.ParDo(TwoTimersDoFn()))

    assert_that(actual_values, equal_to([100]))

    result = p.run()
    result.wait_until_finish()```
I tested it on the master branch and it failed consistently and I get the same error message. I couldn't figure out the root cause but it is caused by two timers for sure. It works fine when there is only one timer used.
I feel I should file a separate jira ticket to address this issue, thoughts?

Reply:
Thanks for the analysis. Filing a JIRA for the multiple-timer issue sounds like the right thing to do here.

Reply:
I checked there is already Jira ticket filed by @tweise (https://issues.apache.org/jira/browse/BEAM-7074)